lib/Sema/DerivedConformanceEquatableHashable.cpp

```diff
@@ -108,6 +132,35 @@ static bool canDeriveConformance(DeclContext *DC,
   return false;
 }
 
+/// Diagnose failed conformance synthesis caused by a member type not conforming
+/// to the same protocol
+void diagnoseFailedDerivation(DeclContext *DC, NominalTypeDecl *nominal,
+                              ProtocolDecl *protocol) {
+  ASTContext &ctx = DC->getASTContext();
+
+  if (auto *enumDecl = dyn_cast<EnumDecl>(nominal)) {
+    auto nonConformingAssociatedTypes =
+        associatedValuesNotConformingToProtocol(DC, enumDecl, protocol);
+    for (auto *typeToDiagnose : nonConformingAssociatedTypes) {
+      ctx.Diags.diagnose(
+          typeToDiagnose->getTypeLoc().getLoc(),
+          diag::missing_member_type_conformance_prevents_synthesis, false,
+          typeToDiagnose->getInterfaceType(), protocol->getDeclaredType());
+    }
+  }
+
+  if (auto *structDecl = dyn_cast<StructDecl>(nominal)) {
+    auto nonConformingStoredProperties =
+        storedPropertiesNotConformingToProtocol(DC, structDecl, protocol);
+    for (auto *propertyToDiagnose : nonConformingStoredProperties) {
+      ctx.Diags.diagnose(
+          propertyToDiagnose->getTypeLoc().getLoc(),
```

Comment:
Stored properties with an inferred type won't have a source location here -- how about just using the source location of the property instead?

Reply:
Good point, I hadn't considered that. fixed